python/lsst/ap/pipe/apPipeTaskRunner.py

```diff
@@ -0,0 +1,152 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+from __future__ import absolute_import, division, print_function
+
+__all__ = ["ApPipeTaskRunner"]
+
+import os
+import sys
+import traceback
+
+import lsst.log
+import lsst.pipe.base as pipeBase
+
+
+class ApPipeTaskRunner(pipeBase.ButlerInitializedTaskRunner):
+
+    def makeTask(self, parsedCmd=None, args=None):
+        """Construct an ApPipeTask with both a Butler and an database.
+
+        Parameters
+        ----------
+        parsedCmd : `argparse.Namespace`
+            Parsed command-line options, as returned by the `~lsst.pipe.base.ArgumentParser`; if specified
+            then args is ignored.
+        args
+            Args tuple passed to `TaskRunner.__call__`. First argument must be
+            the location of the output database and second argument must be a dataref.
+
+        Raises
+        ------
+        RuntimeError
+            Raised if ``parsedCmd`` and ``args`` are both `None`.
+        """
+        if parsedCmd is not None:
+            butler = parsedCmd.butler
+            dbFile = os.path.join(parsedCmd.output, "association.db")
+        elif args is not None:
+            dbFile, dataRef, _ = args
+            butler = dataRef.butlerSubset.butler
+        else:
+            raise RuntimeError("parsedCmd or args must be specified")
+        return self.TaskClass(config=self.config, log=self.log, butler=butler, dbFile=dbFile)
+
+    @staticmethod
+    def getTargetList(parsedCmd, **kwargs):
+        """Get a list of (dbFile, rawRef, calexpRef, kwargs) for `TaskRunner.__call__`.
+        """
+        # Hack to allow makeTask(args). Remove once DM-11767 (or possibly DM-13672) resolved
+        dbFile = os.path.join(parsedCmd.output, "association.db")
+        argDict = dict(
+            templateIds = parsedCmd.templateId.idList,
+            reuse = parsedCmd.reuse,
+            **kwargs
+        )
+        butler = parsedCmd.butler
+        return [(dbFile,
+                 butler.dataRef("raw", **dataId),
+                 dict(calexpRef=butler.dataRef("calexp", **dataId), **argDict))
+                for dataId in parsedCmd.id.idList]
+
+    # TODO: workaround for DM-11767 or DM-13672; can remove once ApPipeTask.__init__ no longer needs dbFile
+    # TODO: find a way to pass the DB argument that doesn't require duplicating TaskRunner.__call__
```

Comment:
No; the former is "remove this workaround once `AssociationTask` handles databases better", the latter is "reduce the code duplication required by this workaround". (And yes, I'm deliberately violating the name-a-ticket rule, because trying to think of a better way to code this isn't worth story points IMO).